File: notebooks/examples/MosaicAndSedona.py
# Databricks notebook source
# MAGIC %md
# MAGIC # Mosaic & Sedona
# MAGIC 
# MAGIC You can combine the usage of Mosaic with other geospatial libraries. 
# MAGIC 
# MAGIC In this example we combine the use of [Sedona](https://sedona.apache.org/setup/databricks/) and Mosaic

# COMMAND ----------

import pyspark.sql.functions as f
from mosaic import enable_mosaic, st_area
from sedona.register.geo_registrator import SedonaRegistrator

enable_mosaic(spark, dbutils)           # Enable Mosaic
SedonaRegistrator.registerAll(spark)    # Register Sedona SQL functions

# COMMAND ----------

df = spark.createDataFrame([{'wkt': 'POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))'}])
(df
   # Mosaic
   .withColumn("mosaic_area", st_area('wkt'))
   # Sedona
   .withColumn("sedona_area", f.expr("ST_Area(ST_GeomFromWKT(wkt))"))
).show()

# COMMAND ----------

spark.createDataFrame([{'wkt': 'POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))'}]).createOrReplaceTempView("sample")

# COMMAND ----------

# MAGIC %md
# MAGIC ## Mosaic & Sedona using SQL

# COMMAND ----------

# MAGIC %scala
# MAGIC import com.databricks.mosaic.functions.MosaicContext
# MAGIC // Register Mosaic in a separate 'mosaic' database, Sedona uses the 'default' database 
# MAGIC MosaicContext.context.register(spark, Some("mosaic"))

# COMMAND ----------

# MAGIC %sql
# MAGIC SELECT 
# MAGIC   mosaic.ST_Area(wkt) as mosaic_area,           -- Mosaic
# MAGIC   ST_Area(ST_GeomFromWKT(wkt)) as sedona_area,  -- Sedona
# MAGIC   wkt
# MAGIC FROM sample

# COMMAND ----------

File: notebooks/examples/scala/MosaicAndSedona.scala
// Databricks notebook source
// MAGIC %md
// MAGIC # Mosaic & Sedona
// MAGIC 
// MAGIC You can combine the usage of Mosaic with other geospatial libraries. 
// MAGIC 
// MAGIC In this example we combine the use of [Sedona](https://sedona.apache.org/setup/databricks/) and Mosaic.
// MAGIC 
// MAGIC ## Setup
// MAGIC 
// MAGIC This notebook will run if you have both Mosaic and Sedona installed on your cluster.
// MAGIC 
// MAGIC ### Install sedona
// MAGIC 
// MAGIC To install Sedona, follow the [official Sedona instructions](https://sedona.apache.org/setup/databricks/).

// COMMAND ----------

// Register Sedona in the 'default' database 
import org.apache.sedona.sql.utils.SedonaSQLRegistrator
SedonaSQLRegistrator.registerAll(spark)

// Import Mosaic functions
import com.databricks.labs.mosaic.functions.MosaicContext
import com.databricks.labs.mosaic.H3
import com.databricks.labs.mosaic.ESRI

val mosaicContext = MosaicContext.build(H3, ESRI)
import mosaicContext.functions._
import org.apache.spark.sql.functions._

// COMMAND ----------

// Example dataset
val df = spark.createDataFrame(Seq(Tuple1("POLYGON ((30 10, 40 40, 20 40, 10 20, 30 10))"))).toDF("wkt")

// COMMAND ----------

df
   .withColumn("mosaic_area", st_area($"wkt"))                      // Mosaic
   .withColumn("sedona_area", expr("ST_Area(ST_GeomFromWKT(wkt))")) // Sedona
   .withColumn("sedona_flipped", expr("ST_FlipCoordinates(ST_GeomFromWKT(wkt))")) // Sedona
   .show()

// COMMAND ----------


